indigo_api/data_migrations/__init__.py

```diff
@@ -94,3 +96,191 @@ def update_namespace(self, xml):
+        "blockList": (re.compile(r"\blist(?P<num>\d+)$"), "list_"),
+        "hcontainer": (re.compile(r"\bcrossheading-(?P<num>\d+)$"), "crossheading_"),
+    }
+
+    complex_replacements = {
+        # e.g. sec_3a.2A.4.1 --> sec_3a.subsec_2A-4-1
+        "subsection": (re.compile(r"^[^.]+(?P<num>(\.\d[\dA-Za-z]*)+)$"), "subsec_"),
+        # e.g. sec_3a.subsec_2-4-1.list_0.b --> sec_3a.subsec_2-4-1.list_0.item_b
+        # or sec_3a.subsec_2-4-1.list_0.3.1 --> sec_3a.subsec_2-4-1.list_0.item_3-1
+        # or sec_4.subsec_4.list_0.item_c.list_0.i --> sec_4.subsec_4.list_0.item_c.list_0.item_i
+        "item": (re.compile(r"^([^.]+\.)+list_\d+(?P<num>(\.[\da-zA-Z]+)+)$"), "item_"),
+    }
+
+    def migrate_act(self, doc):
+        """ Update the xml,
+            - first by updating the values of `id` attributes as necessary,
+            - and then by replacing all `id`s with `eId`s
+        """
+        nsmap = {'a': doc.namespace}
+
+        # basic replacements, e.g. "chapter-" to "chap_"
+        for element, (pattern, replacement) in self.basic_replacements.items():
+            for node in doc.root.xpath(f"//a:{element}", namespaces=nsmap):
+                old_id = node.get("id")
+                new_id = re.sub(pattern, replacement, old_id)
+                rewrite_ids(node, old_id, new_id)
+
+        # add one to n (previously 0-indexed)
+        for element, (pattern, replacement) in self.add_1_replacements.items():
+            for node in doc.root.xpath(f"//a:{element}", namespaces=nsmap):
+                old_id = node.get("id")
+                match = pattern.search(old_id)
+                # e.g. hcontainers aren't all crossheadings
+                if match:
+                    num = int(pattern.search(old_id).group("num")) + 1
+                    prefix = self.get_parent_id(node)
+                    new_id = f"{prefix}.{replacement}{num}"
+                    rewrite_ids(node, old_id, new_id)
+
+        # more complex replacements
+        for element, (pattern, replacement) in self.complex_replacements.items():
+            for node in doc.root.xpath(f"//a:{element}", namespaces=nsmap):
+                old_id = node.get("id")
+                num = self.clean_number(pattern.search(old_id).group("num"))
+                prefix = self.get_parent_id(node)
+                new_id = f"{prefix}.{replacement}{num}"
+                rewrite_ids(node, old_id, new_id)
+
+        # finally
+        # "." separators
+        for node in doc.root.xpath("//a:*[@id]", namespaces=nsmap):
+            old_id = node.get("id")
+            if "." in old_id:
+                node.set("id", old_id.replace(".", "__"))
+
+        # replace all `id`s with `eId`s
+        for node in doc.root.xpath("//a:*[@id]", namespaces=nsmap):
+            node.set("eId", node.get("id"))
+            del node.attrib["id"]
+
+    def clean_number(self, num):
+        return num.strip(".").replace(".", "-")
```

Comment:
This is how slaw does it: https://github.com/longhotsummer/slaw/blob/akn3/lib/slaw/grammars/counters.rb#L37